### ISSUE TITLE
PyTorchClassifier fit function and fit_generator function give different result.

### ISSUE THREAD

**Issue by bettyttytty** [Trusted-AI/adversarial-robustness-toolbox#1217]
**Describe the bug**
I use PyTorchClassifier with model ResNet18 and dataset cifar10. I first use fit function directly without data augmentation to get a promising accuracy (~90%). Since I'd like to incorporate data augmentation, I use PyTorchDataGenerator and fit_generator function to train the model, and I can only get result around 60%.

**To Reproduce**
first load cifar10 data.
```
device = 'cuda' if torch.cuda.is_available() else 'cpu'
(x_train, y_train_onehot), (x_test, y_test_onehot), min_, max_ = load_dataset(str("cifar10"))
x_train = x_train.transpose(0, 3, 1, 2).astype("float32")
x_test = x_test.transpose(0, 3, 1, 2).astype("float32")
y_train_onehot = np.float32(y_train_onehot)
y_test_onehot = np.float32(y_test_onehot)
y_train = np.argmax(y_train_onehot, axis=1)
y_test = np.argmax(y_test_onehot, axis=1)
```
Define Resnet18 model
```
import torch
import torch.nn as nn
import torch.nn.functional as F


class BasicBlock(nn.Module):
    expansion = 1
    def __init__(self, in_planes, planes, stride=1):
        super(BasicBlock, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion * planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion * planes, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion * planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out

class Bottleneck(nn.Module):
    expansion = 4
    def __init__(self, in_planes, planes, stride=1):
        super(Bottleneck, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, self.expansion * planes, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm2d(self.expansion * planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion * planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion * planes, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion * planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = F.relu(self.bn2(self.conv2(out)))
        out = self.bn3(self.conv3(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out

class ResNet(nn.Module):
    def __init__(self, block, num_blocks, num_classes=10):
        super(ResNet, self).__init__()
        self.in_planes = 64

        self.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(64)
        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.linear = nn.Linear(512 * block.expansion, num_classes)

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1] * (num_blocks - 1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * block.expansion
        return nn.Sequential(*layers)

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = F.avg_pool2d(out, 4)
        out = out.view(out.size(0), -1)
        out = self.linear(out)
        return out

def ResNet18():
    return ResNet(BasicBlock, [2, 2, 2, 2])
```
Using fit function:
```
bs = 128
T = 200
lr = 0.1
criterion = nn.CrossEntropyLoss()
cifar_mu = np.ones((3, 32, 32))
cifar_mu[0, :, :] = 0.4914
cifar_mu[1, :, :] = 0.4822
cifar_mu[2, :, :] = 0.4465

cifar_std = np.ones((3, 32, 32))
cifar_std[0, :, :] = 0.2471
cifar_std[1, :, :] = 0.2435
cifar_std[2, :, :] = 0.2616

model = ResNet18().to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr = lr, momentum = 0.9, weight_decay=5e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)
classifier = PyTorchClassifier(model=model, clip_values=(0, 1), loss=criterion, preprocessing=(cifar_mu, cifar_std),
                        optimizer=optimizer, input_shape=(3,32,32), nb_classes=10)

for i in range(T):
    classifier.fit(x_train, y_train, batch_size = bs, nb_epochs=1)
    scheduler.step()
    test_pred = classifier.predict(x_test)
    testacc = np.sum(np.argmax(test_pred, axis=1) == y_test) / len(y_test)
    print('i:{}, test acc:{}%'.format(i, testacc * 100))
    print(classifier.optimizer)
```
this can give a reasonable result. See screenshot:
![image](https://user-images.githubusercontent.com/29892962/124823288-02edba80-df3f-11eb-8031-7a47dd22db61.png)

Using fit_generator function:
```
class CIFAR10_dataset(Dataset):
    def __init__(self, data, targets, transform=None):
        self.data = data
        self.targets = torch.LongTensor(targets)
        self.transform = transform

    def __getitem__(self, index):
        x = Image.fromarray(((self.data[index] * 255).round()).astype(np.uint8).transpose(1, 2, 0))
        x = self.transform(x)
        y = self.targets[index]
        return x, y

    def __len__(self):
        return len(self.data)

bs = 128
T = 200
lr = 0.1
criterion = nn.CrossEntropyLoss()
cifar_mu = np.ones((3, 32, 32))
cifar_mu[0, :, :] = 0.4914
cifar_mu[1, :, :] = 0.4822
cifar_mu[2, :, :] = 0.4465

cifar_std = np.ones((3, 32, 32))
cifar_std[0, :, :] = 0.2471
cifar_std[1, :, :] = 0.2435
cifar_std[2, :, :] = 0.2616

transform = transforms.Compose(
    [transforms.RandomCrop(32, padding=4), 
     transforms.RandomHorizontalFlip(), 
     transforms.ToTensor()]
)
dataset = CIFAR10_dataset(x_train, y_train, transform=transform)
dataloader = DataLoader(dataset, batch_size=bs, shuffle=True)
datagen = PyTorchDataGenerator(iterator=dataloader, size=x_train.shape[0], batch_size=bs)

model = ResNet18().to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr = lr, momentum = 0.9, weight_decay=5e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)
classifier = PyTorchClassifier(model=model, clip_values=(0, 1), loss=criterion, preprocessing=(cifar_mu, cifar_std),
                        optimizer=optimizer, input_shape=(3,32,32), nb_classes=10)

for i in range(T):
    classifier.fit_generator(datagen, nb_epochs=1)
    scheduler.step()
    test_pred = classifier.predict(x_test)
    testacc = np.sum(np.argmax(test_pred, axis=1) == y_test) / len(y_test)
    print('i:{}, test acc:{}%'.format(i, testacc * 100))
    print(classifier.optimizer)
```
I cannot get a good result. See screenshot:
![image](https://user-images.githubusercontent.com/29892962/124823344-13059a00-df3f-11eb-87e7-5d71587cf2ec.png)

**System information (please complete the following information):**
I use google colab to produce the above result.

I'm curious why the result is changing so much. In fact, even if I do not add RandomCrop and RandomHorizontalFlip, but I still use PyTorchDataGenerator to wrap up the data and use fit_generator to train the model, I still cannot get the similar result as directly using the fit function. For the fit_generator usage,I try to use the way that adversarial-robustness-toolbox/examples/adversarial_training_FBF.py did to train the classifier. 

I assume maybe something is wrong with the preprocessing data part but I'm not sure. I'm not exactly sure the difference between using preprocessing in the classifier then directly using fit function, and putting a normalization in the transform so that when calling fit_generator they can get the data from data generator with normalization. I realize in fit_generator there's no apply_preprocessing in fit_generator function but apply_preprocessing exists in fit function, is this a bug?

Please let me know if I did something wrong. Thanks!

**Comment by beat-buesser** [comment]
Hi @bettyttytty Thank you very much for using ART and the detailed code example! I will try it out myself and return soon.

**Comment by beat-buesser** [comment]
Hi @bettyttytty I think your experiment is ok and that you have found a bug. The method 'fit_generator' is applying the preprocessing twice in the fallback for generators with preprocessing. I am fixing it with #1219 and it will be included in ART 1.7.1.

Thank you very much for reporting this issue!

**Comment by bettyttytty** [comment]
Thanks for the reply. Just leave another issue that is related to the preprocessing. I realize in class PyTorchRandomizedSmoothing, we can also apply preprocessing when defining the classifier. But for the fit and predict function, it directly uses the fit and predict function from RandomizedSmoothingMixin, which does not apply the preprocessing step. Please let me know if I miss something here.

**Comment by beat-buesser** [comment]
I think preprocessing is applied correctly because `RandomizedSmoothingMixin.fit` is calling `PyTorchRandomizedSmoothing._fit_classifier` which eventually calls `PyTorchClassifier.fit` where the preprocessing get applied and similar for method `predict.
